Framework/Algorithms/src/ReflectometryBeamStatistics.cpp

```diff
@@ -0,0 +1,458 @@
+          thetaDistribution[i] * pow<2>(massCenter - static_cast<double>(i));
+    }
+    variance /= sum;
+    const double pixelSize = getProperty(Prop::PIXEL_SIZE);
+    rmsVariation =
+        2. * std::sqrt(2. * std::log(2.)) * pixelSize * std::sqrt(variance);
+  }
+  return *rmsVariation;
+}
+
+/** Return true if the sample is considered as bent or beam is divergent.
+ *
+ * @param setup a setup object
+ * @param sampleWaviness the value of sample waviness
+ * @param firstSlitAngularSpread the value of the RMS angular spread of the
+ * first slit
+ * @return true if sample is considered bent
+ */
+bool ReflectometryBeamStatistics::bentSample(
+    const Setup &setup, const double sampleWaviness,
+    const double firstSlitAngularSpread) {
+  return sampleWaviness > 0 &&
+         setup.detectorResolution / setup.l2 > firstSlitAngularSpread;
+}
+
+/** Generate a setup for the reflected beam experiment.
+ *
+ * @param ws the reflectivity workspace
+ * @param directWS corresponding direct beam workspace
+ * @return a setup object
+ * @throw NotFoundError if second slit does not exists
+ */
+const ReflectometryBeamStatistics::Setup
+ReflectometryBeamStatistics::createSetup(const API::MatrixWorkspace &ws,
+                                         const API::MatrixWorkspace &directWS) {
+  Setup s;
+  s.detectorResolution = getProperty(Prop::DETECTOR_RESOLUTION);
+  const std::vector<int> reflectedForeground =
+      getProperty(Prop::REFLECTED_FOREGROUND);
+  auto lowPixel = static_cast<size_t>(reflectedForeground.front());
+  auto highPixel = static_cast<size_t>(reflectedForeground.back());
+  s.foregroundStart = std::min(lowPixel, highPixel);
+  s.foregroundEnd = std::max(lowPixel, highPixel);
+  const std::vector<int> directForeground =
+      getProperty(Prop::DIRECT_FOREGROUND);
+  lowPixel = static_cast<size_t>(directForeground.front());
+  highPixel = static_cast<size_t>(directForeground.back());
+  s.directForegroundStart = std::min(lowPixel, highPixel);
+  s.directForegroundEnd = std::max(lowPixel, highPixel);
+  const auto &spectrumInfo = ws.spectrumInfo();
+  const auto reflectedBeamCentre = static_cast<size_t>(reflectedForeground[1]);
+  s.l2 = spectrumInfo.l2(reflectedBeamCentre);
+  const auto &directSpectrumInfo = directWS.spectrumInfo();
+  const auto directBeamCentre = static_cast<size_t>(directForeground[1]);
+  s.directL2 = directSpectrumInfo.l2(static_cast<size_t>(directBeamCentre));
+  s.pixelSize = getProperty(Prop::PIXEL_SIZE);
+  s.slit1Slit2Distance = interslitDistance(ws);
+  const std::string slit1SizeEntry = getProperty(Prop::FIRST_SLIT_SIZE_LOG);
+  s.slit1Size = slitSize(ws, slit1SizeEntry);
+  s.slit1SizeDirectBeam = slitSize(directWS, slit1SizeEntry);
+  const std::string slit2Name = getProperty(Prop::SECOND_SLIT_NAME);
+  auto instrument = ws.getInstrument();
+  auto slit2 = instrument->getComponentByName(slit2Name);
+  if (!slit2) {
+    throw Kernel::Exception::NotFoundError("Could not find slit", slit2Name);
```

Comment:
I think you check **several times** if the slit component exists for the same workspace ws (interslitDistance calls slitSeparation and there ...). P.S. Shouldn't it be called interSlitDistance?
I suggest input validation. Then you never need to throw for slits.

Reply:
I think it should be interslit, see for example international, internet, etc.